CONTRIBUTING.md

```diff
@@ -17,6 +17,130 @@ limitations under the License.
+4. Set up your development environment according to our guidelines
+
+## Development Environment
+
+### Required Tools
+
+- C++17 compatible compiler
+- Meson build system
+- Ninja build tool
+- clang-format
+- Python (for build scripts and testing)
+- Git with DCO sign-off configured
+
+### Building the Project
+
+NIXL uses Meson and Ninja for building:
+
+```bash
+# Configure the build
+meson setup build
+
+# Build the project
+ninja -C build
+
+# Run tests
+ninja -C build test
+```
+
+### Setting Up clang-format
+
+All C++ code must be formatted using the provided `.clang-format` configuration:
+
+```bash
+# Format a single file
+clang-format -i path/to/file.cpp
+
+# Format all changed files
+git diff --name-only HEAD | grep -E '\.(cpp|h|hpp|cc|cxx)$' | xargs clang-format -i
+```
+
+### Pre-commit Hooks
+
+The project uses pre-commit hooks for Python code quality. Install them with:
+
+```bash
+pip install pre-commit
+pre-commit install
+```
+
+Note: C++ formatting via clang-format in pre-commit is currently optional but recommended.
+
+## Code Standards
+
+### C++17 Guidelines
+
+NIXL is a modern C++17 project. We adhere to the [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines) where appropriate. Key principles:
+
+1. **Use modern C++ features**: Prefer `auto`, range-based loops, structured bindings, `std::optional`, etc.
+2. **RAII everywhere**: Resource management through constructors/destructors
+3. **Smart pointers for ownership**: Use `std::unique_ptr`, `std::shared_ptr` appropriately
+4. **Prefer `const` correctness**: Mark methods and variables `const` when appropriate
+5. **Exception handling**: Exceptions are recommended for control-path code, while error codes should be used for data-path
+
+### STL and Abseil Usage
+
+1. **Prefer STL types**: Use rich STL types as the primary choice
+   - Standard containers: `std::vector`, `std::unordered_map`, etc.
+   - Modern utilities: `std::optional`, `std::variant`, `std::string_view`
+   - Algorithms from `<algorithm>` and `<numeric>`
+
+2. **Fallback to Abseil**: When STL lacks required functionality
+   - String formatting: `absl::StrFormat`
+   - Error handling: `absl::StatusOr` for operations that return values with potential errors
+   - High-performance containers: `absl::flat_hash_map` when needed
+   - Logging utilities (integrated with NIXL logging)
+
+### Code Style
+
+All code must adhere to the style guide in `docs/CodeStyle.md` and be formatted with `.clang-format`. Key points:
+
+1. **Follow the existing code style** in the file/module you're modifying
+2. **Use descriptive names** for variables, functions, and classes
+3. **Document complex logic** with clear comments
+4. **Prefer clarity over cleverness**
+
+### Error Handling
+
+1. **Control-path code**: Use exceptions for exceptional conditions
+   - `std::runtime_error` for runtime failures
+   - `std::invalid_argument` for invalid parameters
+   - Custom exceptions when appropriate
```

Comment:
not sure i agree, maybe we should create more specific exception for Nixl errors and not just throw generic STL exceptions, I'd remove this section for now to progress the PR